### ISSUE TITLE
Generic type no longer narrowed as expected without `extends object`

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#48468]
# Bug Report

### 🕗 Version & Regression Information

- Regression in: 4.7.0-dev.20220325
- Works in 4.7.0-dev.20220323 

### ⏯ Playground Link

Playground currently doesn't show the error:

[Playground link](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220302#code/GYVwdgxgLglg9mABAEwKaoA4FECOICGANgM4A8AKgHwAU+AXIuQDSIBGD5AlA63HIanxIA3gChEExDGCJqUAJ4ZUcGfkQBCALybEAcjisAVqmi7EAH3OIFSlWw3a9B46YtX1ayxtadEYyQGIAE6oUCBBSMBExKgA3OKSAL4JEtKyAIJBQfjyAHQwxJnZ8rS+XkU5+YVZOdQ+vv6BEiFhEYhRJHEpiMkBadQA8kYmULkA1qjyxKW5AmAA5lAAFg46Qy6jE1N1nLOoC8sN3QEt4ZHRXQG9kqdtUEEgXYmIQA)

### 💻 Code

```ts
function deepEquals<T>(a: T, b: T): boolean {
    if (typeof a !== 'object' || typeof b !== 'object' || !a || !b) {
        return false;
    }
    if (Array.isArray(a) || Array.isArray(b)) {
        return false;
    }
    if (Object.keys(a).length !== Object.keys(b).length) { // Error here
        return false;
    }
    return true;
}
```

### 🙁 Actual behavior

The calls to `Object.keys` generate errors:

```
No overload matches this call.
  Overload 1 of 2, '(o: {}): string[]', gave the following error.
    Argument of type 'T' is not assignable to parameter of type '{}'.
  Overload 2 of 2, '(o: object): string[]', gave the following error.
    Argument of type 'T' is not assignable to parameter of type 'object'.
```

### 🙂 Expected behavior

No errors

**Comment by mjbvz** [comment]
This code previously worked for us in VS Code but is now causing around 11 errors in our codebase

**Comment by whzx5byb** [comment]
It seems to be caused by https://github.com/microsoft/TypeScript/pull/48366, which is tagged as "Breaking Change".

**Comment by DanielRosenwasser** [comment]
CC @weswigham 

Looks like the following will work

```ts
function deepEquals<T extends {} | null | undefined>(a: T, b: T): boolean {
    if (typeof a !== 'object' || typeof b !== 'object' || !a || !b) {
        return false;
    }
    if (Array.isArray(a) || Array.isArray(b)) {
        return false;
    }
    if (Object.keys(a).length !== Object.keys(b).length) {
        return false;
    }
    return true;
} 
```

and so will

```ts
function deepEquals<T extends object | null | undefined>(a: T, b: T): boolean {
    if (typeof a !== 'object' || typeof b !== 'object' || !a || !b) {
        return false;
    }
    if (Array.isArray(a) || Array.isArray(b)) {
        return false;
    }
    if (Object.keys(a).length !== Object.keys(b).length) {
        return false;
    }
    return true;
} 
```

but not

```ts
function deepEquals<T extends unknown>(a: T, b: T): boolean {
    if (typeof a !== 'object' || typeof b !== 'object' || !a || !b) {
        return false;
    }
    if (Array.isArray(a) || Array.isArray(b)) {
        return false;
    }
    if (Object.keys(a).length !== Object.keys(b).length) {
        return false;
    }
    return true;
} 
```

Note that the first two only likely work due to @ahejlsberg's change in https://github.com/microsoft/TypeScript/pull/43183. I don't know if we're willing to do something special for `unknown`.

**Comment by DanielRosenwasser** [comment]
Playing around with a fix here FWIW

https://github.com/microsoft/TypeScript/pull/48576

**Comment by arnold-O** [comment]
The Generic type without constraint is vague, hence with the **extends** keyword you clearly state what you want that Generic type to be, which is the reason for **Typescript**.
